File: frida_scripts/templates/jni_discovery.js
/**
 * JNI Discovery Script - JNI 함수 동적 탐지 스크립트
 *
 * ## 목적:
 * Android 앱의 JNI 함수와 RegisterNatives 호출을 런타임에 탐지합니다.
 *
 * ## 탐지 대상:
 * 1. **RegisterNatives 호출**: Dynamic JNI 매핑 (Java 메서드 → Native 함수)
 * 2. **Native 메서드 열거**: Reflection으로 native 키워드가 있는 메서드 탐색
 *
 * ## JNI 브릿지 종류:
 * - **Static JNI**: Java_com_example_ClassName_methodName 이름 규칙
 * - **Dynamic JNI**: RegisterNatives로 임의의 함수 이름 매핑 (난독화에 사용)
 *
 * ## 출력 정보:
 * - Java 클래스 및 메서드 이름
 * - JNI 시그니처 (예: "(I)V")
 * - Native 함수 주소 (메모리 주소)
 */

Java.perform(function() {
    console.log("[*] JNI Discovery started");

    var jniMethods = [];  // 열거로 발견한 native 메서드
    var registerNativesCalls = [];  // RegisterNatives 호출 목록

    // RegisterNatives 후킹 - Dynamic JNI 매핑 가로채기
    // JNI_OnLoad에서 호출되는 RegisterNatives를 후킹하여 매핑 정보 추출
    try {
        var env = Java.vm.getEnv();  // JNIEnv 포인터 획득
        var RegisterNatives = env.registerNatives;

        if (RegisterNatives) {
            Interceptor.attach(RegisterNatives.implementation, {
                onEnter: function(args) {
                    try {
                        // RegisterNatives 함수 시그니처:
                        // jint RegisterNatives(JNIEnv* env, jclass clazz,
                        //                      const JNINativeMethod* methods, jint nMethods)
                        //
                        // args[0] = JNIEnv* (환경 포인터)
                        // args[1] = jclass (Java 클래스 객체)
                        // args[2] = JNINativeMethod* (메서드 배열 포인터)
                        // args[3] = jint (등록할 메서드 개수)

                        var className = null;
                        var methods = [];
                        var nMethods = args[3].toInt32();

                        // Java 클래스 이름 추출
                        try {
                            var jclass = Java.cast(args[1], Java.use("java.lang.Class"));
                            className = jclass.getName();
                        } catch (e) {
                            className = "<unknown>";
                        }

                        // JNINativeMethod 배열 파싱
                        // C 구조체를 메모리에서 직접 읽어 정보 추출
                        if (nMethods > 0 && nMethods < 100) {  // 유효성 검사 (너무 큰 값 방지)
                            var methodsArray = args[2];

                            for (var i = 0; i < nMethods; i++) {
                                try {
                                    // JNINativeMethod 구조체 레이아웃:
                                    // struct {
                                    //     const char* name;       // +0 (메서드 이름)
                                    //     const char* signature;  // +4/+8 (JNI 시그니처)
                                    //     void* fnPtr;            // +8/+16 (함수 포인터)
                                    // };
                                    //
                                    // 32비트: 각 포인터 4바이트 → 구조체 크기 12바이트
                                    // 64비트: 각 포인터 8바이트 → 구조체 크기 24바이트

                                    var offset = Process.pointerSize === 8 ? i * 24 : i * 12;
                                    var methodStruct = methodsArray.add(offset);

                                    // 구조체 필드 읽기
                                    var namePtr = methodStruct.readPointer();  // name 포인터 읽기
                                    var sigPtr = methodStruct.add(Process.pointerSize).readPointer();  // signature
                                    var fnPtr = methodStruct.add(Process.pointerSize * 2).readPointer();  // fnPtr

                                    // C 문자열 읽기
                                    var methodName = namePtr.readCString();
                                    var methodSig = sigPtr.readCString();

                                    methods.push({
                                        name: methodName,
                                        signature: methodSig,
                                        address: fnPtr.toString()
                                    });
                                } catch (e) {
                                    console.log("[-] Error reading method " + i + ": " + e);
                                }
                            }
                        }

                        // Python으로 결과 전송
                        var registrationInfo = {
                            type: 'jni_register',
                            className: className,
                            methodCount: nMethods,
                            methods: methods,
                            timestamp: new Date().toISOString()
                        };

                        registerNativesCalls.push(registrationInfo);
                        send(registrationInfo);  // Python의 on_message 핸들러로 전송

                        console.log("[+] RegisterNatives: " + className + " (" + nMethods + " methods)");
                        methods.forEach(function(m) {
                            console.log("    - " + m.name + m.signature + " @ " + m.address);
                        });

                    } catch (e) {
                        console.log("[-] Error in RegisterNatives hook: " + e);
                    }
                }
            });

            console.log("[+] RegisterNatives hook installed");
        }
    } catch (e) {
        console.log("[-] Failed to hook RegisterNatives: " + e);
    }

    // Enumerate all loaded classes and find native methods
    function enumerateNativeMethods() {
        console.log("[*] Enumerating native methods...");

        Java.enumerateLoadedClasses({
            onMatch: function(className) {
                try {
                    var clazz = Java.use(className);

                    // Get all declared methods
                    var methods = clazz.class.getDeclaredMethods();

                    methods.forEach(function(method) {
                        try {
                            var modifiers = method.getModifiers();
                            // Check if method is native (0x100 = NATIVE modifier)
                            if ((modifiers & 0x100) !== 0) {
                                var methodInfo = {
                                    type: 'native_method',
                                    className: className,
                                    methodName: method.getName(),
                                    signature: method.toString(),
                                    modifiers: modifiers.toString(16)
                                };

                                jniMethods.push(methodInfo);

                                console.log("[+] Native method: " + className + "." + method.getName());
                            }
                        } catch (e) {
                            // Skip methods that throw exceptions
                        }
                    });

                } catch (e) {
                    // Skip classes that cannot be loaded
                }
            },
            onComplete: function() {
                console.log("[*] Native method enumeration complete");
                console.log("[*] Found " + jniMethods.length + " native methods");

                send({
                    type: 'jni_summary',
                    nativeMethods: jniMethods,
                    registerNativesCalls: registerNativesCalls,
                    totalNativeMethods: jniMethods.length,
                    totalRegistrations: registerNativesCalls.length
                });
            }
        });
    }

    // Start enumeration after a short delay
    setTimeout(enumerateNativeMethods, 2000);

    console.log("[*] JNI Discovery initialized");
});

File: unat/discovery/jni_discovery.py
#!/usr/bin/env python3
"""
JNI Discovery Module
Discovers JNI functions and RegisterNatives calls in Android applications
"""

import logging
import time
from typing import List, Dict, Optional
from dataclasses import dataclass, field
from pathlib import Path

try:
    import frida
    FRIDA_AVAILABLE = True
except ImportError:
    FRIDA_AVAILABLE = False
    logging.warning("Frida not available. JNI discovery features will be limited.")


@dataclass
class JNINativeMethod:
    """Data class to store JNI native method information"""
    class_name: str
    method_name: str
    signature: str
    address: Optional[str] = None
    is_registered: bool = False  # True if found via RegisterNatives


@dataclass
class JNIRegistration:
    """Data class to store RegisterNatives call information"""
    class_name: str
    method_count: int
    methods: List[Dict[str, str]] = field(default_factory=list)
    timestamp: str = ""


@dataclass
class JNIDiscoveryResult:
    """Data class to store JNI discovery results"""
    package_name: str
    total_native_methods: int = 0
    total_registrations: int = 0
    native_methods: List[JNINativeMethod] = field(default_factory=list)
    registrations: List[JNIRegistration] = field(default_factory=list)


class JNIDiscovery:
    """
    JNI Discovery for finding JNI functions and RegisterNatives calls
    Uses Frida to dynamically inspect JNI bridge
    """

    def __init__(self, frida_device, package_name: str):
        """
        Initialize JNI Discovery

        Args:
            frida_device: Frida device object
            package_name: Target package name

        Raises:
            ImportError: If Frida is not available
        """
        if not FRIDA_AVAILABLE:
            raise ImportError("Frida is required but not installed")

        self.logger = logging.getLogger(__name__)
        self.device = frida_device
        self.package_name = package_name
        self.session: Optional[frida.core.Session] = None
        self.script: Optional[frida.core.Script] = None

        # Storage for discovered data
        self.native_methods: List[JNINativeMethod] = []
        self.registrations: List[JNIRegistration] = []

    def attach(self, spawn: bool = False) -> bool:
        """
        Attach to the target process

        Args:
            spawn: If True, spawn the app; if False, attach to running process

        Returns:
            bool: True if attached successfully
        """
        try:
            if spawn:
                self.logger.info(f"Spawning {self.package_name}...")
                pid = self.device.spawn([self.package_name])
                self.session = self.device.attach(pid)
                self.device.resume(pid)
            else:
                self.logger.info(f"Attaching to {self.package_name}...")
                self.session = self.device.attach(self.package_name)

            self.logger.info("Attached successfully")
            return True

        except frida.ProcessNotFoundError:
            self.logger.error(f"Process not found: {self.package_name}")
            return False
        except Exception as e:
            self.logger.error(f"Failed to attach: {e}")
            return False

    def detach(self):
        """Detach from the target process"""
        if self.script:
            self.script.unload()
            self.script = None

        if self.session:
            self.session.detach()
            self.session = None

        self.logger.info("Detached")

    def discover(self, timeout: int = 5) -> JNIDiscoveryResult:
        """
        Perform JNI discovery

        Args:
            timeout: How long to wait for discovery (seconds)

        Returns:
            JNIDiscoveryResult: Discovery results
        """
        if not self.session:
            raise RuntimeError("Not attached to any process. Call attach() first.")

        # Load JNI discovery script
        script_path = Path(__file__).parent.parent.parent / "frida_scripts" / "templates" / "jni_discovery.js"

        if script_path.exists():
            with open(script_path, 'r', encoding='utf-8') as f:
                script_code = f.read()
        else:
            # Fallback: inline script (simplified version)
            script_code = self._get_jni_discovery_script()

        self.logger.info("Starting JNI discovery...")

        # Execute script
        script = self.session.create_script(script_code)

        def on_message(message, data):
            if message['type'] == 'send':
                payload = message['payload']
                msg_type = payload.get('type')

                if msg_type == 'jni_register':
                    # RegisterNatives call detected
                    registration = JNIRegistration(
                        class_name=payload.get('className', '<unknown>'),
                        method_count=payload.get('methodCount', 0),
                        methods=payload.get('methods', []),
                        timestamp=payload.get('timestamp', '')
                    )
                    self.registrations.append(registration)

                    # Also add to native methods list
                    for method in payload.get('methods', []):
                        native_method = JNINativeMethod(
                            class_name=payload.get('className', '<unknown>'),
                            method_name=method.get('name', ''),
                            signature=method.get('signature', ''),
                            address=method.get('address', ''),
                            is_registered=True
                        )
                        self.native_methods.append(native_method)

                elif msg_type == 'native_method':
                    # Native method found via enumeration
                    native_method = JNINativeMethod(
                        class_name=payload.get('className', ''),
                        method_name=payload.get('methodName', ''),
                        signature=payload.get('signature', ''),
                        is_registered=False
                    )
                    self.native_methods.append(native_method)

                elif msg_type == 'jni_summary':
                    # Final summary
                    self.logger.info(
                        f"JNI Discovery complete: {payload.get('totalNativeMethods', 0)} "
                        f"native methods, {payload.get('totalRegistrations', 0)} registrations"
                    )

            elif message['type'] == 'error':
                self.logger.error(f"Script error: {message.get('description', 'Unknown error')}")

        script.on('message', on_message)
        script.load()

        # Wait for discovery to complete
        self.logger.info(f"Waiting {timeout} seconds for JNI discovery...")
        time.sleep(timeout)

        script.unload()

        # Build result
        result = JNIDiscoveryResult(
            package_name=self.package_name,
            total_native_methods=len(self.native_methods),
            total_registrations=len(self.registrations),
            native_methods=self.native_methods,
            registrations=self.registrations
        )

        return result

    @staticmethod
    def _get_jni_discovery_script() -> str:
        """Fallback inline script for JNI discovery"""
        return """
        Java.perform(function() {
            console.log("[*] JNI Discovery (fallback) started");

            // Enumerate native methods
            Java.enumerateLoadedClasses({
                onMatch: function(className) {
                    try {
                        var clazz = Java.use(className);
                        var methods = clazz.class.getDeclaredMethods();

                        methods.forEach(function(method) {
                            try {
                                var modifiers = method.getModifiers();
                                if ((modifiers & 0x100) !== 0) {
                                    send({
                                        type: 'native_method',
                                        className: className,
                                        methodName: method.getName(),
                                        signature: method.toString()
                                    });
                                }
                            } catch (e) {}
                        });
                    } catch (e) {}
                },
                onComplete: function() {
                    console.log("[*] Native method enumeration complete");
                }
            });
        });
        """


def discover_jni_functions(
    frida_device,
    package_name: str,
    spawn: bool = False,
    timeout: int = 5
) -> Optional[JNIDiscoveryResult]:
    """
    Convenience function to discover JNI functions

    Args:
        frida_device: Frida device object
        package_name: Target package name
        spawn: Spawn the app instead of attaching
        timeout: Discovery timeout in seconds

    Returns:
        Optional[JNIDiscoveryResult]: Discovery results or None if failed
    """
    discovery = JNIDiscovery(frida_device, package_name)

    try:
        if not discovery.attach(spawn=spawn):
            return None

        result = discovery.discover(timeout=timeout)
        return result

    except Exception as e:
        logging.error(f"JNI discovery failed: {e}")
        return None

    finally:
        discovery.detach()
